accepted/future-releases/class-modifiers/feature-specification.md

```diff
@@ -631,127 +624,82 @@ It is a compile-time error to:
-restriction applies even to types within the same library.
+restriction applies even to types within the same library._
 
-*Once the ability to use as an interface is removed, it cannot be reintroduced
+_Once the ability to use as an interface is removed, it cannot be reintroduced
 in a subtype. If a class is marked `base` or `final`, you may still implement
 the class's interface inside the same library, but the implementing class must
 again be marked `base`, `final`, or `sealed` to avoid it exposing an
-implementable interface.*
+implementable interface._
 
 Further, while you can ignore some restrictions on declarations within the same
 library, you cannot use that to ignore restrictions inherited from other
 libraries.
 
-We say that `S` is a _direct declared superinterface_ of a class, mixin, or
-mixin class declaration `D` if `D` has a superclass clause of the form
-`C with M1 .. Mk` (where `k` may be zero when there is no `with` clause)
-and `S` is `C`, or `S` is `Mj` for some `j` in 1 .. k,
-or if `D` has an `implements` or `on` clause and `S` occurs as one of the operands of
-that clause.
-
-We then say that a class or mixin declaration `D` *cannot be implemented locally* if it
-has a direct declared superinterface `S` such that:
-
-*   `S` is from another library than `D`, and `S` has the modifier `base`,
-    `final` or `sealed`, or
-
-    ```dart
-    // a.dart
-    base mixin class S {}
-
-    // b.dart
-    import 'a.dart';
-
-    // These cannot be implemented locally:
-    sealed class DE extends S {}
-    final class DM with S {}
-    base mixin MO on S {}
-    ```
-
-*   `S` is from the same library as `D`, and `S` cannot be implemented locally.
-
-    ```dart
-    // a.dart
-    base class B {}
-
-    // b.dart
-    import 'a.dart';
-
-    // These cannot be implemented locally (from the previous rule):
-    base class S extends B {}
-    base mixin M on B {}
+We say that an interface *cannot be implemented locally* in a library *L*
+if it has any superinterface *S* whose declaration is marked `base` or `final`,
+and *S* is in a library other than *L*.
 
-    // And thus these also cannot be implemented locally:
-    base class DE extends S {}
-    base class DM extends B with M {} // (from this and the previous rule).
-    base mixin MO on S {}
-    base mixin M2 on M {}
-    ```
-
-Otherwise, `D` can be implemented locally.
-
-It is a compile-time error if:
-
-*   A class, mixin, or mixin class declaration `D` has an `implements` clause
-    where `S` is an operand, and `S` is a class, mixin, or mixin class
-    declaration declared in the same library as `D`, and `S` cannot be
-    implemented locally.
-
-    ```dart
-    // a.dart
-    base class B {}
+It's a compile-time error if a `class`, `mixin`, `mixin class` or `enum`
+declaration in library *L* has an `implements` clause with an entry
+denoting the interface *S*, and *S* prevents implementation locally in *L*.
```

Comment:
I think by "prevents implementation locally" you mean "cannot be implemented locally"?  Or do you meant to change the definition above to define "prevents implementation locally"?

Reply:
If there's a superinterface *S* whose declaration is marked `base` or `final`, then according to the *prevents implementation* rule, every subtype of *S* will also be marked `base` or `final`.

Given that, I think it's sufficient to say:

> It's a compile-time error if a `class`, `mixin`, `mixin class` or `enum` declaration has an `implements` clause with an entry referring to a type declared in another library and that type is marked `base` or `final`.

I don't think we need to define any transitive "cannot be implemented locally" rule at all. The fact that `base` or `final` is already required to percolate down the entire subtype graph means we can rely on it being in any immediate supertype.